Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -42,15 +45,93 @@ public static void WarnAboutSuspiciousDependenciesForImplementation(DafnyOptions
     if (results.Any(r => r.Outcome != SolverOutcome.Valid)) {
       return;
     }
+    var distilled = results.Select(r => (r.CoveredElements, DafnyConsolePrinter.DistillVCResult(r))).ToList();
+    var asserts = distilled.Select(tp => (tp.CoveredElements, new HashSet<(Boogie.IToken Tok, string Description, string Id)>(tp.Item2.Asserts))).ToList();
+    var unusedFunctions = UnusedFunctions(depManager, name, distilled.SelectMany(r => r.CoveredElements), distilled.SelectMany(r => r.Item2.AvailableAxioms));
+    Warn(dafnyOptions, reporter, depManager, name, asserts, unusedFunctions);
+  }
+
+  private static List<Function> UnusedFunctions(ProofDependencyManager depManager, string name, IEnumerable<TrackedNodeComponent> coveredElements, IEnumerable<Axiom> axioms) {
+    var unusedFunctions = new List<Function>();
+    if (depManager.idsByMemberName[name].Decl is Method m) {
+      var referencedFunctions = ReferencedFunctions(m);
+
+      var hiddenFunctions = HiddenFunctions(referencedFunctions);
+
+      var usedFunctions = coveredElements.Select(depManager.GetFullIdDependency).OfType<FunctionDefinitionDependency>()
+        .Select(dep => dep.function).Deduplicate((a, b) => a.Equals(b));
+
+      unusedFunctions = referencedFunctions.Except(hiddenFunctions).Except(usedFunctions).ToList();
+
+      HashSet<Function> ReferencedFunctions(Method method) {
```

Comment:
Please only use inner functions when you need access to local variables. Basically, try to avoid them.

Reply:
Why?

Reply:
Local functions _can_ use variables from the containing method, so you don't know its inputs by only looking at its signature, making it harder to figure out what it does. This one doesn't use anything from the outer method which you can signify by moving it to the class scope.

Secondly, when you do use local functions it's often better to place them at the end of the method.